providers/ldap/ldap_client.go

```diff
@@ -0,0 +1,655 @@
+	return l.searchLdap(query, c.GroupScope)
+}
+
+func (l *LClient) searchLdap(query string, scope string) ([]client.Identity, error) {
+	c := l.ConstantsConfig
+	searchConfig := l.SearchConfig
+	identities := []client.Identity{}
+	var search *ldap.SearchRequest
+
+	searchDomain := l.Config.Domain
+	if strings.EqualFold(c.UserScope, scope) {
+		search = ldap.NewSearchRequest(searchDomain,
+			ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+			query,
+			searchConfig.UserSearchAttributes, nil)
+	} else {
+		if l.Config.GroupSearchDomain != "" {
+			searchDomain = l.Config.GroupSearchDomain
+		}
+		search = ldap.NewSearchRequest(searchDomain,
+			ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+			query,
+			searchConfig.GroupSeachAttributes, nil)
+	}
+
+	lConn, err := l.newConn()
+	if err != nil {
+		return []client.Identity{}, fmt.Errorf("Error %v creating connection", err)
+	}
+	defer lConn.Close()
+	results, err := lConn.Search(search)
+	if err != nil {
+		ldapErr, ok := reflect.ValueOf(err).Interface().(*ldap.Error)
+		if ok && ldapErr.ResultCode != ldap.LDAPResultNoSuchObject {
+			return []client.Identity{}, fmt.Errorf("When searching ldap from /v1/identity Failed to search: %s, error: %#v", query, err)
+		}
+	}
+
+	for i := 0; i < len(results.Entries); i++ {
+		entry := results.Entries[i]
+		identity, err := l.attributesToIdentity(entry.Attributes, results.Entries[i].DN, scope)
+		if err != nil {
+			return []client.Identity{}, err
+		}
+		identities = append(identities, *identity)
+	}
+
+	return identities, nil
+}
+
+func escapeLDAPSearchFilter(filter string) string {
+	buf := new(bytes.Buffer)
+	for i := 0; i < len(filter); i++ {
+		currChar := filter[i]
+		switch currChar {
+		case '\\':
+			buf.WriteString("\\5c")
+			break
+		case '*':
+			buf.WriteString("\\2a")
+			break
+		case '(':
+			buf.WriteString("\\28")
+			break
+		case ')':
+			buf.WriteString("\\29")
+			break
+		case '\u0000':
+			buf.WriteString("\\00")
+			break
+		default:
+			buf.WriteString(string(currChar))
+		}
+	}
+	return buf.String()
+}
+
+func (l *LClient) TestLogin(testAuthConfig *model.TestAuthConfig) error {
+	log.Info("Now generating Ldap token")
+	split := strings.Split(testAuthConfig.Code, ":")
+	username, password := split[0], split[1]
+	externalId := username
+	if testAuthConfig.AuthConfig.LdapConfig.LoginDomain != "" {
+		externalId = testAuthConfig.AuthConfig.LdapConfig.LoginDomain + "\\" + username
+	}
+
+	if password == "" {
+		return fmt.Errorf("Failed to login, password not provided")
+	}
+
+	ldapServer := testAuthConfig.AuthConfig.LdapConfig.Server
+	ldapPort := testAuthConfig.AuthConfig.LdapConfig.Port
+	log.Info("Now creating Ldap connection")
+	lConn, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", ldapServer, ldapPort))
+	if err != nil {
+		return fmt.Errorf("Error %v creating connecting", err)
+	}
+	lConn.SetTimeout(time.Duration(l.Config.ConnectionTimeout) * time.Second)
```

Comment:
Use the testConfig. ConnectionTimeout, if it is not specified then use the l.Config.ConnectionTimeout